### PR TITLE
Add build configs/builds to overview details.

### PR DESCRIPTION
**Pull request by TheRealJon**
image

@openshift/team-ux-review

### CODE REVIEW

**Comment by TheRealJon** [comment]
@spadgett @beanh66 @bmignano @cshinn Let me know what you think about having that 'Start Build' action next to the Build Config. I put it there for myself while testing and though it might be good to just keep there.

**Comment by spadgett** [comment]
> @spadgett @beanh66 @bmignano @cshinn Let me know what you think about having that 'Start Build' action next to the Build Config. I put it there for myself while testing and though it might be good to just keep there.

+1, we had a start build action in the previous overview

File: frontend/public/components/build.tsx [hunk 1]
```diff
@@ -42,6 +43,22 @@ export enum BuildStrategyType {
   Source = 'Source',
 }
 
+export const BuildPhaseIcon = ({build}) => {
+  const {status: {phase}} = build;
+  switch (phase) {
+    case 'Running':
+      return <span className="fa fa-spin fa-refresh" />;
```
Comment (spadgett) [inline frontend/public/components/build.tsx]:
```suggestion
      return <span className="fa fa-spin fa-refresh" aria-hidden="true" />;
```

File: frontend/public/components/overview/build-configs-overview.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import { Link } from 'react-router-dom';
+import { Button, ListGroup } from 'patternfly-react';
+
+import { BuildPhaseIcon } from '../build';
+import { errorModal } from '../modals/error-modal';
+import { startBuild } from '../../module/k8s/builds';
+import { fromNow } from '../utils/datetime';
+import { ResourceLink, resourcePath, SidebarSectionHeading } from '../utils';
+
+const conjugateBuildPhase = (phase) => {
```
Comment (spadgett) [inline frontend/public/components/overview/build-configs-overview.tsx]:
```suggestion
const conjugateBuildPhase = (phase: string) => {
```

File: frontend/public/components/overview/build-configs-overview.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    case 'Failed':
+      return 'failed';
+    default:
+      return `is ${_.toLower(phase)}`;
+  }
+};
+
+const BuildOverviewItem = ({build}) => {
+  const {metadata} = build;
+  const timestamp = build.status.completionTimestamp || build.status.startTimestamp;
```
Comment (spadgett) [inline frontend/public/components/overview/build-configs-overview.tsx]: I think these could both be unset for new builds that haven't started running. (Not uncommon for builds with serial policy.)

File: frontend/public/components/overview/build-configs-overview.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    </li>
+    {
+      _.isEmpty(builds)
+        ? <li className="list-group-item"><span className="text-muted">No Builds found for this Build Config</span></li>
+        : _.map(builds, build => <BuildOverviewItem key={build.metadata.uid} build={build} />)
+    }
+  </ListGroup>;
+};
+
+export const BuildConfigsOverview: React.SFC<any> = ({buildConfigs}) => <div className="build-config-overview">
```
Comment (spadgett) [inline frontend/public/components/overview/build-configs-overview.tsx]: Exported components generally should have props types defined. (Ideally the other components above, too.)

File: frontend/public/components/overview/index.tsx [hunk 5]
```diff
@@ -182,6 +193,26 @@ const sortReplicationControllersByRevision = (replicationControllers: K8sResourc
   return sortByRevision(replicationControllers, getDeploymentConfigVersion);
 };
 
+const sortBuilds = (builds) => {
```
Comment (spadgett) [inline frontend/public/components/overview/index.tsx]:
```suggestion
const sortBuilds = (builds: K8sResourceKind[]) => {
```

File: frontend/public/components/overview/index.tsx [hunk 6]
```diff
@@ -486,11 +521,45 @@ class OverviewDetails extends SafetyFirst<OverviewDetailsProps, OverviewDetailsS
     };
   }
 
+  getBuildsForResource(buildConfig) {
```
Comment (spadgett) [inline frontend/public/components/overview/index.tsx]:
```suggestion
  getBuildsForResource(buildConfig: K8sResourceKind) {
```

File: frontend/public/components/overview/index.tsx [hunk 7]
```diff
@@ -486,11 +521,45 @@ class OverviewDetails extends SafetyFirst<OverviewDetailsProps, OverviewDetailsS
...
+    const {builds} = this.props;
+    return getOwnedResources(buildConfig, builds.data);
+  }
+
+  getBuildConfigsForResource(resource) {
```
Comment (spadgett) [inline frontend/public/components/overview/index.tsx]:
```suggestion
  getBuildConfigsForResource(resource: K8sResourceKind) {
```

File: frontend/public/components/overview/index.tsx [hunk 8]
```diff
@@ -486,11 +521,45 @@ class OverviewDetails extends SafetyFirst<OverviewDetailsProps, OverviewDetailsS
...
+    const {buildConfigs} = this.props;
+    const nativeTriggers = _.get(resource, 'spec.triggers');
+    const annotatedTriggers = JSON.parse(_.get(resource, ['metadata','annotations', TRIGGERS_ANNOTATION], '[]'));
```
Comment (spadgett) [inline frontend/public/components/overview/index.tsx]: We should catch exceptions parsing the JSON, or we white screen if someone edits the value to something bad.

```ts
const getAnnotatedTriggers = (obj: K8sResourceKind) => {
  const triggersJSON = getAnnotation(resource, TRIGGERS_ANNOTATION) || '[]';
  try {
    return JSON.parse(triggersJSON);
  } catch(e) {
    console.warn('Error parsing triggers annotation', e);
    return [];
  }
};
```

We should eventually pull this into a utility that we can use on workload details pages, but we can do that later.

File: frontend/public/components/overview/index.tsx [hunk 9]
```diff
@@ -486,11 +521,45 @@ class OverviewDetails extends SafetyFirst<OverviewDetailsProps, OverviewDetailsS
...
+    const triggers = _.unionWith(nativeTriggers, annotatedTriggers, _.isEqual);
+    return _.flatMap(triggers, (trigger) => {
+      const from = trigger.from || _.get(trigger, 'imageChangeParams.from', {});
+      if ( from.kind !== 'ImageStreamTag') {
+        return [];
+      }
+      return _.reduce(buildConfigs.data, (acc, buildConfig) => {
+        if (_.get(buildConfig, 'spec.output.to.name') === from.name) {
```
Comment (spadgett) [inline frontend/public/components/overview/index.tsx]: I think we need to check the namespace, too. IIRC, it is possible for a build in one namespace to push to an image stream in another namespace.

File: frontend/public/components/overview/build-configs-overview.tsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+          namespace={metadata.namespace}
+        />
+      </div>
+      <div>
+        <Button bsStyle="link" onClick={onClick}>Start Build</Button>
+      </div>
+    </li>
+    {
+      _.isEmpty(builds)
+        ? <li className="list-group-item"><span className="text-muted">No Builds found for this Build Config</span></li>
```
Comment (jhadvig) [inline frontend/public/components/overview/build-configs-overview.tsx]: nit: missing period at the end of the sentence

**Comment by beanh66** [comment]
Agreed @TheRealJon I think it looks great and the addition of the action seems helpful! A few questions, and apologies if you are already addressing these:

1. Are you currently showing all the builds, or at what point would build 1 disappear in your example
2. We talked about only showing the number instead of the number appended to the name since it will just be the same repeating name as the build config above.  You can see an example in the documentation [here](https://github.com/openshift/openshift-origin-design/blob/master/web-console/4.0-designs/overview/overview.md#resources-tab) Thoughts? 
3. Will you also be showing a bit of the log snippet on the overview or only allow linking to the full log view?

**Comment by cshinn** [comment]
> > @spadgett @beanh66 @bmignano @cshinn Let me know what you think about having that 'Start Build' action next to the Build Config. I put it there for myself while testing and though it might be good to just keep there.
> 
> +1, we had a start build action in the previous overview

I think something like a `btn-default btn-xs` might work better in this case than a link without being too obtrusive

**Comment by TheRealJon** [comment]
@beanh66
> 1. Are you currently showing all the builds, or at what point would build 1 disappear in your example

I am currently showing all of the builds, I missed the part of the design spec that says it should be limited to 3. Should we also provide a link to the builds list page when we are truncating the list?

> 2. We talked about only showing the number instead of the number appended to the name since it will just be the same repeating name as the build config above.  You can see an example in the documentation [here](https://github.com/openshift/openshift-origin-design/blob/master/web-console/4.0-designs/overview/overview.md#resources-tab) Thoughts?

That's feasible, I think I was looking at the wrong example in the google doc when I created this. There are two different screen shots in there that show the builds and I mixed them up.

> 3. Will you also be showing a bit of the log snippet on the overview or only allow linking to the full log view?

For this PR, it just has the link to the full log view. I will be working on adding the log snippet in a follow on.

**Comment by spadgett** [comment]
> I am currently showing all of the builds, I missed the part of the design spec that says it should be limited to 3. Should we also provide a link to the builds list page when we are truncating the list?

The build config link does have a builds tab, which should contain a filtered list. If we add a link, we probably want to link there. Although maybe the existing build config link is sufficient since it's just one click on the tab to see all builds.

Note that by default pruning is enabled for build configs, but if it's disabled by the user, this list could get pretty long. We do have more space with the scrolling pane than we did before.

Either way, I think you're usually interested in the most recent finished (complete or failed) and anything pending or running. You should also do some testing with serial run policy set in the build config as well.

**Comment by TheRealJon** [comment]
@beanh66 How should we handle the edge case where there are 3 or more builds queued and one running. I would think we will always want to show the running build, but that creates a strange scenario for sort order. Do we show 2 of the new/pending builds and the one running? What about other completed or failed builds? Those will get pushed off the list by new builds, but user's might be interested in seeing the last build that was completed/failed/cancelled. Maybe we show only the one new build, one running build, and the most recent terminated build (completed, failed, or cancelled)?

**Comment by TheRealJon** [comment]
> I think you're usually interested in the most recent finished (complete or failed) and anything pending or running. You should also do some testing with serial run policy set in the build config as well.

I did test with serial policy and started a bunch of new builds. Basically, you end up with a long list of new builds, with the currently running build below them. I'm thinking instead of showing all new builds, we only show the one that will run next, and then some sort of indication of the number of new builds that are queued.

File: frontend/public/components/overview/index.tsx (hunk 9 above)
Comment (TheRealJon) [reply frontend/public/components/overview/index.tsx]: So just check that the buildconfig and resource are in the same namespace?

File: frontend/public/components/overview/index.tsx (hunk 8 above)
Comment (TheRealJon) [reply frontend/public/components/overview/index.tsx]: Yeah, I feel like a lot of the functions we have in the top of this file could be pulled out into utilities.

File: frontend/public/components/overview/build-configs-overview.tsx (hunk 3 above)
Comment (TheRealJon) [reply frontend/public/components/overview/build-configs-overview.tsx]: Should we add the creation time also? That way new builds will also have a relative time shown.

Comment (spadgett) [reply frontend/public/components/overview/build-configs-overview.tsx]: > Should we add the creation time also? That way new builds will also have a relative time shown.

Yeah, I think so.

Comment (spadgett) [reply frontend/public/components/overview/build-configs-overview.tsx]: Maybe change the variable name to `lastUpdated`?

**Comment by beanh66** [comment]
@TheRealJon That makes sense to me. I agree showing all could get lengthy. Definitely always showing running builds makes sense too. @bmignano and I were trying to do something similar to what we had in the old console, that seemed to work well. Thanks for the info on (2) and (3) above as well. Sorry for the change for (2) we originally did have all the names listed out but realized how repetitive it was. Thanks!

**Comment by TheRealJon** [comment]
/hold

Need to address the way build names are displayed in the list. Using generic build number instead of repetitive build name.

**Comment by TheRealJon** [comment]
@spadgett I think this is ready for final review/merging

**Comment by spadgett** [comment]
We should create follow-on issues for only showing recent builds and adding build logs to the overview. Thanks @TheRealJon !

**Pull request merged**